ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron.hs

```diff
@@ -927,45 +1039,147 @@ instance (ByronGiven, Typeable cfg, ConfigContainsGenesis cfg, NoUnexpectedThunk
-applyByronGenTx validationMode (ByronEBBLedgerConfig (ByronLedgerConfig cfg)) genTx (ByronEBBLedgerState st@ByronLedgerState{..}) =
-    (\x -> ByronEBBLedgerState $ st { blsCurrent = x }) <$> go genTx blsCurrent
+applyByronGenTx validationMode
+                (ByronEBBLedgerConfig (ByronLedgerConfig cfg))
+                genTx
+                (ByronEBBLedgerState st@ByronLedgerState{blsCurrent}) =
+    (\x -> ByronEBBLedgerState $ st { blsCurrent = x })
+      <$> go genTx blsCurrent
   where
-    go :: GenTx (ByronBlockOrEBB cfg)
+    go :: (MonadError ByronApplyTxError m)
+       => GenTx (ByronBlockOrEBB cfg)
        -> CC.Block.ChainValidationState
-       -> Except CC.UTxO.UTxOValidationError CC.Block.ChainValidationState
-    go (ByronTx { byronTx = tx }) cvs = wrapCVS <$>
-        runReaderT (CC.UTxO.updateUTxO env utxo [tx]) validationMode
+       -> m CC.Block.ChainValidationState
+    go genTx cvs = case genTx of
+        ByronTx             _ tx       -> applyByronTx tx
+        ByronDlg            _ cert     -> applyByronDlg cert
+        ByronUpdateProposal _ proposal -> applyByronUpdateProposal proposal
+        ByronUpdateVote     _ vote     -> applyByronUpdateVote vote
       where
-        wrapCVS newUTxO = cvs { CC.Block.cvsUtxo = newUTxO }
-        protocolMagic = fixPM $ CC.Genesis.configProtocolMagic cfg
+        protocolMagic = fixPM (CC.Genesis.configProtocolMagic cfg)
+
+        k = CC.Genesis.configK cfg
+
+        currentEpoch = CC.Slot.slotNumberEpoch (CC.Genesis.configEpochSlots cfg)
+                                              currentSlot
+
+        currentSlot = CC.Block.cvsLastSlot cvs
+
         utxo = CC.Block.cvsUtxo cvs
+
+        dlgState = CC.Block.cvsDelegationState cvs
+
         updateState = CC.Block.cvsUpdateState cvs
-        env = CC.UTxO.Environment
-          { CC.UTxO.protocolMagic = protocolMagic
-          , CC.UTxO.protocolParameters = CC.UPI.adoptedProtocolParameters updateState
-          , CC.UTxO.utxoConfiguration = CC.Genesis.configUTxOConfiguration cfg
-          }
-        fixPM (Crypto.AProtocolMagic a b) = Crypto.AProtocolMagic (reAnnotate a) b
-
-mkByronTx :: CC.UTxO.ATxAux ByteString -> GenTx (ByronBlockOrEBB cfg)
-mkByronTx tx = ByronTx
-    { byronTx   = tx
-    , byronTxId = Crypto.hash $ CC.UTxO.taTx tx
-      -- TODO replace this with a function from cardano-ledger, see
-      -- cardano-ledger#581
-    }
+
+        delegationMap =
+          (V.Interface.delegationMap . CC.Block.cvsDelegationState) cvs
+
+        numGenKeys = toNumGenKeys $ Set.size (allowedDelegators cfg)
+
+        toNumGenKeys :: Integral n => n -> Word8
+        toNumGenKeys n
+          | n > fromIntegral (maxBound :: Word8) = error $
+            "toNumGenKeys: Too many genesis keys"
+          | otherwise = fromIntegral n
+
+        fixPM (Crypto.AProtocolMagic a b) =
+          Crypto.AProtocolMagic (reAnnotate a) b
+
+        applyByronTx tx = wrapCVS <$>
```

Comment:
This refactoring is much better :+1: 

Next step: I'm wondering whether it would make sense to define all these `env`s in the shared `where` block and to rename them to `utxoEnv`, `interfaceEnv`, etc. This will allow you to share the `upiEnv` across the two cases and it will instantly make clear that these `env`s have different types (now the four `env`s have three different types). Similarly for `wrapCVS`, e.g. `wrapUTxO`, `wrapDlgState`, ... What do you think?